internal/controller/application_controller.go

```diff
@@ -163,49 +177,134 @@ func (r *ApplicationPullServiceAccountCreator) getImageRepositoryPullSecretNames
+			}
+		} else {
+			log.Info("Skipping secret due to unexpected type (only kubernetes.io/dockerconfigjson is processed)", "secretName", secretName, "type", secret.Type)
 		}
-	} else {
-		return nil
 	}
 
-	// Create service account for application
-	secretsReferences := []corev1.ObjectReference{}
-	imagePullSecretsReferences := []corev1.LocalObjectReference{}
-	for _, secretName := range pullSecretNames {
-		secretsReferences = append(secretsReferences, corev1.ObjectReference{Name: secretName})
-		imagePullSecretsReferences = append(imagePullSecretsReferences, corev1.LocalObjectReference{Name: secretName})
+	// Marshal combined auths back into .dockerconfigjson format
+	combinedDockerConfig := dockerConfigJson{Auths: combinedAuths}
+	marshaledData, err := json.Marshal(combinedDockerConfig)
+	if err != nil {
+		log.Error(err, "failed to marshal combined docker config json")
+		return nil, err
 	}
 
-	applicationSA := corev1.ServiceAccount{
+	// Create the application pull secret
+	applicationPullSecret = &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      serviceAccountName,
+			Name:      applicationPullSecretName,
 			Namespace: application.Namespace,
 		},
-		Secrets:          secretsReferences,
-		ImagePullSecrets: imagePullSecretsReferences,
+		Type: corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: marshaledData,
+		},
+	}
+
+	// Set pull secret ownership to application
+	if err := controllerutil.SetOwnerReference(application, applicationPullSecret, r.Scheme); err != nil {
+		log.Error(err, "failed to set application as owner of application pull secret", "applicationName", application.Name)
+		return nil, err
 	}
 
-	// set serviceAccount ownership to application
-	if err := controllerutil.SetOwnerReference(application, &applicationSA, r.Scheme); err != nil {
-		log.Error(err, "failed to set application as owner of SA", "applicationName", application.Name)
-		return err
+	if err := r.Client.Create(ctx, applicationPullSecret); err != nil {
+		log.Error(err, "failed to create application pull secret", "secretName", applicationPullSecretName, l.Action, l.ActionAdd)
+		return nil, err
 	}
 
-	if err := r.Client.Create(ctx, &applicationSA); err != nil {
-		log.Error(err, "failed to create service account", "serviceAccountName", l.Action, l.ActionAdd)
-		return err
+	return applicationPullSecret, nil
+}
+
+// udateServiceAccountWithApplicationPullSecret updates the ServiceAccount to include
+// the application pull secret as an imagePullSecret and as a Secret
+func (r *ApplicationPullSecretCreator) updateServiceAccountWithApplicationPullSecret(ctx context.Context, sa *corev1.ServiceAccount, applicationPullSecret *corev1.Secret) error {
+	log := ctrllog.FromContext(ctx)
+
+	// Check and update imagePullSecrets
+	foundPull := false
+	for _, pullSecret := range sa.ImagePullSecrets {
+		if pullSecret.Name == applicationPullSecret.Name {
+			foundPull = true
```

Comment:
instead foundPull foundSecret
just have secretLinked

and add it there if it is False, AND have new var shouldUpdateSA

Reply:
updated